tests/test_symlink.py

```diff
@@ -55,10 +86,130 @@ def test_files_are_visible_in_symlink_folder(symlink_config: pathlib.Path):
+        e.relative_to(root_path) for e in expected_paths if e not in expected_before_symlink_paths
+    ]
     assert (
         not files_added_by_symlinking
     ), f"Symlinking created files or folders in read/only space {files_added_by_symlinking}"
     # Check that the symlink files are there
-    missing_files = [e for e in expected_paths if e not in found]
+    missing_files = [e.relative_to(root_path) for e in expected_paths if e not in found]
     assert not missing_files, f"There were missing files: {missing_files}\n found: {found}\n expected: {expected_paths}"
+    extra_files = [pathlib.Path(e).relative_to(root_path) for e in found if e not in expected_paths]
+    assert not extra_files, f"There were extra files: {extra_files}\n found: {found}\n expected: {expected_paths}"
+    return out
+
+
+@pytest.fixture
+def s3_endpoint_url(monkeypatch):
+    """Uses moto to start a mocked S3 endpoint on a local port"""
+    monkeypatch.setenv("AWS_ACCESS_KEY_ID", "testing")
+    monkeypatch.setenv("AWS_SECRET_ACCESS_KEY", "testing")
+    monkeypatch.setenv("AWS_SECURITY_TOKEN", "testing")
+    monkeypatch.setenv("AWS_SESSION_TOKEN", "testing")
+
+    port = 7000
+    endpoint_url = f"http://127.0.0.1:{port}"
+    started_server = False
+    i = 0
+    # try ports from 7000 to 8000 for a valid one
+    # Finds an open port to start the server on
+    for i in range(1000):
+        # First: check that the port is closed (curl should fail or we skip to the next port)
+        try:
+            subprocess.check_output(["curl", endpoint_url], timeout=5)
+            continue
+        except:
+            pass
+        # Start the S3 service
+        server = ThreadedMotoServer(port=port + i)
+        server.start()
+        endpoint_url = f"http://127.0.0.1:{port + i}"
+        # CHeck that the server is working
+        try:
+            subprocess.check_output(["curl", endpoint_url], timeout=5)
+            started_server = True
+            break
+        except:
+            server.stop()
+            continue
+    if not started_server:
+        raise ValueError(f"Failed to mock S3 server on ports {port} to {port + i}")
+    else:
+        print(f"Started Mock S3 server at {endpoint_url}")
+    yield endpoint_url
+    server.stop()
+
+
+@pytest.fixture
+def s3_datasets(symlink_config: pathlib.Path, s3_endpoint_url: str):
+    """Uploads the mocked datasets to a mock S3"""
+    bucket = "sdk"
+    conn = boto3.resource("s3", endpoint_url=s3_endpoint_url)
+    try:
+        conn.create_bucket(
+            Bucket=bucket,
+            CreateBucketConfiguration={"LocationConstraint": s3_endpoint_url},
+        )
+    except Exception as error:
+        if "BucketAlreadyOwnedByYou" in str(error):
+            pass
+        else:
+            raise
+
+    symlink_def: Dict[str, List[str]] = json.loads(symlink_config.read_text())
+    new_symlink_def = {}
+    sources = [pathlib.Path(source) for sources in symlink_def.values() for source in sources]
```

Comment:
I think we're redefining `sources` to be the values of the `symlink_def` dict below, so this line could be removed?